libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +177,65 @@ namespace rttr { namespace mapGenerator {
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            mineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        mineableResourceInfo mRIs[4] = {mineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        mineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        mineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        mineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite)};
         int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
 
         RTTR_FOREACH_PT(MapPoint, map.size)
         {
             if(textures.All(pt, IsMinableMountain))
             {
+                int ratio = 0;
                 int randomNumber = rnd.RandomValue(1, total);
-                int ratio = settings.ratioGold;
-
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Gold + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioCoal;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Coal + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioIron;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Iron + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioGranite;
-                if(randomNumber < ratio)
+                for(auto mRI : mRIs)
                 {
-                    resources[pt] = libsiedler2::R_Granite + rnd.RandomValue(0, 8);
+                    ratio += mRI.ratio;
+
+                    // Wrong resource, look further.
+                    if(randomNumber > ratio)
+                        continue;
+
+                    // Right resource, adjust and check the budget.
+                    ++mRI.budget;
+                    if(mRI.budget <= 0)
+                        break;
+
+                    // Budget is positive, avoid overwriting existing clusters though
+                    if(resources[pt])
+                        continue;
+
+                    // Cluster the resource around the map point.
+                    for(int xd = -4; xd <= 4; ++xd)
+                    {
+                        for(int yd = -4; yd <= 4; ++yd)
+                        {
+                            MapPoint pt_d(pt.x + xd, pt.y + yd);
```

Comment:
I'd suggest to use either MakeMapPoint or `GetPointsInRadiusWithCenter` (maybe base on https://github.com/Return-To-The-Roots/s25client/pull/1373 which changed that a bit) Currently you are basically ignoring the wrap-around. The latter will make these patches round instead of square, your choice which you like better.
I'm also curious about the `-4`: Wouldn't that always yield a point where resources have been placed already?

Reply:
round is fine, -4 is explained above, will change this.